[week 7] 인가(Authorization) 구현하기

pjkfckr opened the pull request:
### 목표
급한 마음을 가지고 망치기 보단 다시작성했던 코드도 뒤돌아보며 참고할수 있는 좋은 코드 작성하기



이번주 도 잘부탁드립니다.

pjkfckr committed:
UserController와 ProductController 테스트 진행

johngrib commented on app/src/main/java/com/codesoom/assignment/advice/NotFoundErrorAdvice.java:
```diff
@@ -0,0 +1,25 @@
+    @ResponseBody
+    @ResponseStatus(HttpStatus.NOT_FOUND)
+    @ExceptionHandler(UserNotFoundException.class)
+    public ErrorResponse handleUserNotFound() { return new ErrorResponse("User not found"); }
```
이렇게 한 줄로 join하지 않고 중괄호가 있는 곳에서 개행을 해주면 좋겠습니다.
```suggestion
    public ErrorResponse handleUserNotFound() {
        return new ErrorResponse("User not found");
    }
```

pjkfckr replied:
`printwidth` 가 설정이 너무 길게 되어있어서 자꾸만 개행을 해도 변경이 되었는데 이번에 옵션을 찾아서 변경했습니다!

johngrib reviewed (commented):
커밋 하나에 변경된 파일 68개가 들어있어요. 이렇게 커밋 하나의 사이즈가 크면 동료가 리뷰하기 어렵습니다. 다음 커밋부터는 주제별로 분리해서 커밋을 해보도록 합시다.

johngrib commented on app/src/test/java/com/codesoom/assignment/domain/UserTest.java:
```diff
@@ -1,52 +1,26 @@
-
-        assertThat(user.authenticate("test")).isFalse();
-        assertThat(user.authenticate("xxx")).isFalse();
+    @DisplayName("of() 메소드로 User 생성할 때")
```
문장을 완성하는 것이 좋습니다.

pjkfckr replied:
문장을 완성하여 `of() 메소드로 User 객체를 생성할 수 있다` 로 변경하였습니다!

pjkfckr committed:
UserTest의 문장 완성 및 중괄호 개행으로 변경

pjkfckr committed:
Jwt 발급을 위한 JwtUtil 클래스 구현

pjkfckr committed:
Token의 유효성 검사 후 발생하는 Custom Exception 추가

pjkfckr committed:
사용자를 찾아서 로그인 하는 login메소드 구현

pjkfckr committed:
email로 로그인하는 테스트 구현

pjkfckr committed:
로그인 시 password 포함 테스트

pjkfckr committed:
로그인 할 수 없는 이메일 테스트

pjkfckr committed:
사용자를 생성하는 로직을 하나의 메소드로 분리

pjkfckr committed:
비밀번호가 다를때 테스트

pjkfckr commented:
commit 을 쪼개는 것이 자신을 위해서도 좋은 방향이지만 자꾸만 습관이 안들어서 통으로 보냈습니다.
앞으로는 최대한 하나의 기능을 하나의 commit 하는 방향으로 고치도록 하겠습니다.

johngrib commented on app/src/main/java/com/codesoom/assignment/utils/JwtUtil.java:
```diff
@@ -0,0 +1,46 @@
+                    .setSigningKey(key)
+                    .build()
+                    .parseClaimsJws(token)
+                    .getBody();
+        } catch (SignatureException e) {
+            throw new AuthException();
+        }
```
stacktrace를 출력할 필요는 없을까요?

pjkfckr replied:
`stacktrace`가 아닌 유효하지않은 토큰 이라는것을 알려주는 `Custom Exception`으로 변경하는것이 좋을것 같습니다.

johngrib reviewed (commented):
이번주부터는 PR을 여러개로 나누어 작업해보도록 합시다. 상영님만 괜찮다면 이 PR은 `CodeSoom:pjkfckr`에 머지하고 다음 작업부터는 작업 주제별로 별도의 PR로 올려보도록 하죠. 어떻게 해야 할 지 잘 모르겠으면 질문을 달아주시면 됩니다.

pjkfckr commented:
PR을 여러개 나누는 것이 origin/pjkfckr 에서 기능 단위로 branch를 따로 나누는 것 일까요?

pjkfckr committed:
JWT토큰 유효성 테스트

pjkfckr committed:
로그인 하는 SessionController 테스트

johngrib commented:
> PR을 여러개 나누는 것이 origin/pjkfckr 에서 기능 단위로 branch를 따로 나누는 것 일까요?

맞습니다. PR을 보내는 타깃은 `CodeSoom:pjkfckr` 로 하고, 계속 새로운 브랜치를 만들어서 여기로 PR을 보내는 거죠. OK 하시면 바로 이 PR을 머지하도록 하겠습니다. 기능별로 이름을 잘 붙인 PR을 보내주시면 돼요. 여러개 올리셔도 됩니다. 모두 제가 리뷰해드릴 거고요.

johngrib commented on app/src/main/java/com/codesoom/assignment/application/authentication/AuthenticationService.java:
```diff
@@ -36,6 +36,11 @@ public String login(String email, String password) {
+    /**
+     * 생성한 토큰에서 claims에 있는 id 파싱
+     * @param accessToken Claim에 id가 담겨있는 토큰
+     * @return claims에서 id를 리턴
```
이렇게 해도 괜찮겠죠.
`@return 토큰에 포함된 id`

pjkfckr replied:
좋은거 같습니다!

johngrib commented on app/src/main/java/com/codesoom/assignment/controller/SessionController.java:
```diff
@@ -0,0 +1,41 @@
+    public SessionResponseDto login(
+            @RequestBody @Valid SessionRequestDto sessionRequestDto
+    ) {
+        String email = sessionRequestDto.getEmail();
+        String password = sessionRequestDto.getPassword();
+
+        String accessToken = authenticationService.login(email, password);
```
이런 곳에는 `final`을 잘 붙여주는 것도 의미있는 연습이라 할 수 있을 것 같아요.

pjkfckr replied:
수정할 수 없는 변수로 지정하여 혹시나 나중에 다른곳에서 변경되는 것을 막기위해 `final`을 붙이는 것이 좋은 것 같습니다!

pjkfckr commented:
머지 부탁드립니다!

pjkfckr committed:
PR 리뷰 수정

SessionController의 login메소드 변수에 final 붙이기
AuthenticationService의 parseToken메소드 javaDoc return 설명 수정
